Add button to eject iso images from instance console

edlerd opened the pull request:
## Done

- added button to eject iso images from instance console

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @lorumic or @edlerd for access.
    - With a local copy of this branch, run as described [here](../HACKING.md#setting-up-for-development).
2. Perform the following QA steps:
    - launch instance with an iso, go to console start instance, check eject button is there and works.

webteam-app commented:
Demo starting at https://lxd-ui-455.demos.haus

lorumic commented on src/pages/instances/actions/EjectIsoBtn.tsx:
```diff
@@ -0,0 +1,51 @@
+      .catch((e: Error) => notify.failure("Eject ISO failed", e));
+  };
+
+  if (!attachedIso) {
+    return null;
+  }
```
I think this can be moved before the `ejectIso` function.

lorumic reviewed (commented):
What I don't like here is that the action is not reversible. If I eject an ISO and want to "put it back" where it was, I need to manually edit the YAML, which is not something accessible for the inexperienced user.

In other words, we're adding a "user-friendly" button to eject the ISO, but then we leave the inexperienced user with no way to "roll back" to the previous state (instance with an ISO attached).

Can we improve the feature in this sense? For example, we could use the same "Eject ISO" button also as an "Attach ISO" button, hence showing it all the time for VM instances. Then we branch on the `attachedIso` variable: if it's undefined, we show the "Attach ISO" version, allowing the user to add an ISO volume (under the hood, this modifies the `devices` object of the instance), otherwise we use the logic that is already implemented to eject the ISO.

By doing this, we're essentially creating a virtual CD/DVD reader for Windows VMs - isn't that cool?

edlerd commented:
> What I don't like here is that the action is not reversible. If I eject an ISO and want to "put it back" where it was, I need to manually edit the YAML, which is not something accessible for the inexperienced user.

A valid concern and an interesting idea to have an "Attach ISO" button in case there is no ISO on the VM. Though, I am not sure if that is confusing for the most part of our users that are "just running" VMs. I'll discuss this later today with @piperdeck

lorumic commented:
> > What I don't like here is that the action is not reversible. If I eject an ISO and want to "put it back" where it was, I need to manually edit the YAML, which is not something accessible for the inexperienced user.
> 
> A valid concern and an interesting idea to have an "Attach ISO" button in case there is no ISO on the VM. Though, I am not sure if that is confusing for the most part of our users that are "just running" VMs. I'll discuss this later today with @piperdeck

I think that if they can grasp the concept of "ejecting" an ISO, they can easily grasp also the other way round. So I don't see it more confusing than the feature that is being added here.

edlerd commented:
> I think that if they can grasp the concept of "ejecting" an ISO, they can easily grasp also the other way round. So I don't see it more confusing than the feature that is being added here.

The eject is visible to users that want to install from an ISO. Most users will run a VM from an image server and will never see the eject button. My concern was to avoid polluting the interface with an action that is not relevant in the current context. However, writing this now we can distinguish both cases, because the ISO VMs will have no base image. So we can limit the attach new ISO button to the VMs that were launched with an ISO.

lorumic commented:
> The eject is visible to users that want to install from an ISO. Most users will run a VM from an image server and will never see the eject button. My concern was to avoid polluting the interface with an action that is not relevant in the current context. However, writing this now we can distinguish both cases, because the ISO VMs will have no base image. So we can limit the attach new ISO button to the VMs that were launched with an ISO.

Personally I would add it to all VMs, because I see it as the missing "virtual CD/DVD reader" feature that competitor products have (one example [here](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vm_admin.doc/GUID-C58B93A7-52CF-456D-95C1-8B5A906C9619.html)).

Now, if we're talking about how to avoid "polluting" the interface, that's a completely different topic because I think the interface is polluted already for a different reason: button labels. Which brings me to an "old" idea that I mentioned in the past: a "control bar" (see the comment [here](https://github.com/canonical/lxd-ui/pull/383#issuecomment-1585124378)) could be a good place to contain this kind of feature buttons, which should eventually be converted to _icon_ buttons.

edlerd commented:
Putting back in draft mode, because this is based on #445 hence appearing to have more changes than just the single commit it adds.

piperdeck reviewed (changes requested):
Overall, I would like more icons to make these buttons on the console screen more identifiable at a glance, but I realize we don't actually have the right assets to make that happen. For now, I would suggest putting the ISO file name next to the eject button, instead of in a tooltip. The first reason is that it makes it easier to quickly identify the ISO, and the second reason is that it provides a visual contrast between the two states of having or not having an ISO attached.

![Frame 4 (2)](https://github.com/canonical/lxd-ui/assets/45367207/70e8309f-deb5-433e-97a6-83913fd9ecf3)

edlerd commented:
Thanks @piperdeck that looks much better indeed. I updated this PR to use the muted text instead of the tooltip next to the eject button. 

@lorumic I also updated from the upstream changes, so it is now ready for code review as well.

piperdeck reviewed (approved):
Looks great!

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,120 @@
+          </>
+        )
+      )
+      .catch((e: Error) => notify.failure("Detach ISO failed", e))
```
Is there a reason why we use "eject" as the verb for the success message but "detach" here? I would use a bit of consistency and change this to:
```suggestion
      .catch((e: Error) => notify.failure("ISO ejection failed", e))
```

edlerd replied:
We had a discussion about it and said it would be consistent to use detach/attach in more places of the UI.

lorumic replied:
Ok, then maybe it would be worth removing all the "eject"s? We have a mixed usage at the moment.

edlerd replied:
I think the latest state of the branch removed them all.

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,120 @@
+  return attachedIso ? (
+    <>
+      <span className="u-text--muted margin-right">{attachedIso.source}</span>
+      <SubmitButton
```
If it's about having a button that shows a loader while an operation is running, the most semantically correct choice here would probably be react-component's [`ActionButton`](https://canonical.github.io/react-components/?path=/docs/actionbutton--default-story).

edlerd replied:
Good point!

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,120 @@
+        appearance=""
+        isSubmitting={isLoading}
+        isDisabled={false}
+        buttonLabel="Detach ISO"
```
Here too, I suggest using the verb "eject", as it is probably more familiar in this kind of context, but if there's a particular reason you chose "detach", feel free to disregard this suggestion.
```suggestion
        buttonLabel="Eject ISO"
```

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,120 @@
+  const handleSelect = (image: RemoteImage) => {
+    setLoading(true);
+    closePortal();
+    const values = getInstanceEditValues(instance);
+    values.devices.push({
+      type: "iso-volume",
+      name: "iso-volume",
+      bare: {
+        "boot.priority": "10",
+        pool: image.pool ?? "",
+        source: image.aliases,
+        type: "disk",
+      },
+    });
```
This chunk of code is currently used in 2 different places - here and in `CreateInstanceForm`. Maybe it would be good to create a reusable helper function for it?

edlerd replied:
Very good point! I wanted to extract this but forgot -- will do that now!

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,120 @@
+    </>
+  ) : (
+    <>
+      <SubmitButton
```
Same as above - I suggest using `ActionButton` here.

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,114 @@
+    <>
+      <span className="u-text--muted margin-right">{attachedIso.source}</span>
+      <ActionButton
+        appearance=""
+        loading={isLoading}
+        disabled={false}
```
I think 2 props here are unneeded, since they're specifying the value that is already the default.
```suggestion
        loading={isLoading}
```

edlerd replied:
Ah right, I missed than when switching the component. Thanks :)

lorumic commented on src/pages/instances/actions/AttachIsoBtn.tsx:
```diff
@@ -0,0 +1,114 @@
+  ) : (
+    <>
+      <ActionButton
+        appearance=""
+        loading={isLoading}
+        disabled={false}
```
Same as above:
```suggestion
        loading={isLoading}
```
